src/Microsoft.Graph.Core/Tasks/PageIterator.cs

```diff
@@ -230,14 +233,50 @@ public async Task ResumeAsync()
         /// </summary>
         /// <param name="token">The CancellationToken used to stop iterating calls for more pages.</param>
         /// <returns>The task object that represents the results of this asynchronous operation.</returns>
-        /// <exception cref="Microsoft.CSharp.RuntimeBinder.RuntimeBinderException">Thrown when a base CollectionPage that does not implement NextPageRequest
-        /// is provided to the PageIterator</exception>
         /// <exception cref="Microsoft.Graph.ServiceException">Thrown when the service encounters an error with
         /// a request.</exception>
         public async Task ResumeAsync(CancellationToken token)
         {
             await IterateAsync(token);
         }
+
+        /// <summary>
+        /// Helper method to extract the collection rom an <see cref="IParsable"/> instance.
+        /// </summary>
+        /// <param name="parsableCollection">The <see cref="IParsable"/> to extract the collection from</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentException">Thrown when the object doesn't contain a collection inside it</exception>
+        private static List<TEntity> ExtractEntityListFromParsable(IParsable parsableCollection)
+        {
+            return parsableCollection.GetType().GetProperty("Value").GetValue(parsableCollection, null) as List<TEntity> ?? throw new ArgumentException("The Parsable does not contain a collection property");
+        }
+
+        /// <summary>
+        /// Helper method to execute the the request to get back the next page
+        /// </summary>
+        /// <param name="nextPageRequestInformation">The <see cref="RequestInformation"/> of the next page request</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to use to cancel requests</param>
+        /// <returns></returns>
+        private async Task<IParsable> GetNextPageAsync(RequestInformation nextPageRequestInformation, CancellationToken cancellationToken)
+        {
+            // Use reflection to invoke the "requestAdapter.SendAsync" method so as to be able to be able to pass the generic type arguments
+            // neededed for handling the response.
+            var method = _client.RequestAdapter.GetType().GetMethod("SendAsync").MakeGenericMethod(_currentPage.GetType());
+            var sendTask = (Task)method.Invoke(_client.RequestAdapter, new object[] { nextPageRequestInformation, null, cancellationToken });
```

Comment:
technically this is a task of the same type you get line 53. So by adding a second generic parameter to the class, with a constraint where U : IParsable, and reusing that generic type line 53 and here, you could avoid having to use reflection to get the result.

Reply:
Thats true.
I was mainly at a bit of a crossroads between doing this versus the reflection route as this means that the developer's experience is altered in the way that creating a Page Iterator changes from this

```cs
            eventPageIterator = PageIterator<Event>.CreatePageIterator(baseClient, originalPage, (e) =>
            {
                events.Add(e);
                return true;
            });
```

to this (where the developer has the extra work to specify the collection type and I couldn't tell if it was worth it from their perspective, so I was trying to save the hassle). However, from our perspective I really appreciate the type safety.

```cs
eventPageIterator = PageIterator<Event, EventsResponse>.CreatePageIterator(baseClient, originalPage, (e) =>
            {
                events.Add(e);
                return true;
            });
```

This is now resolved in the latest update.